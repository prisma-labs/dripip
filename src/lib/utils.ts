import { inspect } from 'util'

export function dump(arg: unknown): void {
  console.error(inspect(arg, { depth: 30 }))
}
/**
 * Indent a string that spans multiple lines.
 */
export function indentBlock(size: number, block: string): string {
  return block
    .split('\n')
    .map((line) => range(size).map(constant(' ')).join('') + line)
    .join('\n')
}

export const indentBlock4 = indentBlock.bind(null, 4)

/**
 * Create a function that will only ever return the given value when called.
 */
export const constant = <T>(x: T): (() => T) => {
  return function () {
    return x
  }
}

/**
 * Create a range of integers.
 */
const range = (times: number): number[] => {
  const list: number[] = []
  while (list.length < times) {
    list.push(list.length + 1)
  }
  return list
}

// type IndexableKeyTypes = string | number | symbol

// type Indexable<T = unknown> = Record<string | number, T>

// type JustIndexableTypes<T> = T extends IndexableKeyTypes ? T : never

// type KeysMatching<Rec, Keys> = NonNullable<
//   {
//     [RecKey in keyof Rec]: Rec[RecKey] extends Keys ? RecKey : never
//   }[keyof Rec]
// >

// export type GroupBy<T extends Indexable, K extends IndexableKeys<T>> = {
//   [KV in JustIndexableTypes<T[K]>]?: Array<T extends Record<K, KV> ? T : never>
// }

// type IndexableKeys<Rec> = KeysMatching<Rec, IndexableKeyTypes>

// export function groupByProp<
//   Obj extends Indexable,
//   KeyName extends IndexableKeys<Obj>
// >(xs: Obj[], keyName: KeyName): GroupBy<Obj, KeyName> {
//   type KeyValue = JustIndexableTypes<Obj[KeyName]>
//   const seed = {} as GroupBy<Obj, KeyName>

//   return xs.reduce((groupings, x) => {
//     const groupName = x[keyName] as KeyValue

//     if (groupings[groupName] === undefined) {
//       groupings[groupName] = []
//     }

//     // We know the group will exist, given above initializer.
//     groupings[groupName]!.push(
//       x as Obj extends Record<KeyName, KeyValue> ? Obj : never
//     )

//     return groupings
//   }, seed)
// }

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
export function casesHandled(x: never): never {
  throw new Error(`A case was not handled for value: ${x}`)
}

/**
 * Determine if the given array or object is empty.
 */
export function isEmpty(x: {} | unknown[]): boolean {
  return Array.isArray(x) ? x.length === 0 : Object.keys(x).length > 0
}

/**
 * Pause in time for given milliseconds.
 */
export function delay(milliseconds: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, 1000)
  })
}

/**
 * Like Array.findIndex but working backwards from end of array.
 */
export function findIndexFromEnd<T>(xs: T[], f: (x: T) => boolean): number {
  for (let index = xs.length - 1; index > -1; index--) {
    if (f(xs[index])) return index
  }
  return -1
}

/**
 * Get the last element of an array.
 */
export function last<T>(xs: T[]): null | T {
  if (xs.length === 0) return null
  return xs[xs.length - 1]
}

export function numericAscending(n1: number, n2: number): -1 | 0 | 1 {
  if (n1 < n2) return -1
  if (n1 > n2) return 1
  return 0
}
