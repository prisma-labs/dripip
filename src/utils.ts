import * as Git from 'simple-git/promise'
import * as Semver from 'semver'

// const emptyCommit = (git:Git.SimpleGit, options: string[]): Promise<Git.CommitSummary> => {
//   git.commit(["--allow-empty", "--message", "initial commit"])
// }

const getReleaseTagsAtCommit = (ref: string): Promise<Semver.SemVer[]> =>
  Git()
    .tag({ '--points-at': ref })
    .then(result =>
      result
        .trim()
        .split('\n')
        .map(tag => Semver.parse(tag))
        .filter((parsed): parsed is Semver.SemVer => parsed !== null)
    )

const indentBlock4 = (block: string): string => indentBlock(4, block)

const indentBlock = (size: number, block: string): string => {
  return block
    .split('\n')
    .map(
      line =>
        range(size)
          .map(constant(' '))
          .join('') + line
    )
    .join('\n')
}

const constant = <T>(x: T): (() => T) => {
  return function() {
    return x
  }
}

const range = (times: number): number[] => {
  const list: number[] = []
  while (list.length < times) {
    list.push(list.length + 1)
  }
  return list
}

export { getReleaseTagsAtCommit, indentBlock, indentBlock4 }
