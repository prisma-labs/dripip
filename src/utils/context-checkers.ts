import { validateNPMAuthSetup } from '../lib/npm-auth'
import { Validator } from './context-guard'

export function isTrunk(): Validator {
  return {
    code: 'must_be_on_trunk',
    summary: 'You must be on the trunk branch',
    run(ctx) {
      return ctx.currentBranch.isTrunk
    },
  }
}

export function branchSynced(): Validator {
  return {
    code: 'branch_not_synced_with_remote',
    summary: 'Your branch must be synced with the remote',
    run(ctx) {
      if (ctx.currentBranch.syncStatus === 'synced') {
        return true
      } else {
        return {
          kind: 'fail',
          details: {
            syncStatus: ctx.currentBranch.syncStatus,
          },
        }
      }
    },
  }
}

export function npmAuthSetup(): Validator {
  return {
    code: 'npm_auth_not_setup',
    summary: 'You must have npm auth setup to publish to the registry',
    run() {
      const result = validateNPMAuthSetup()

      if (result.kind === 'pass') {
        return true
      } else {
        return { kind: 'fail', details: { reasons: result.reasons } }
      }
    },
  }
}
