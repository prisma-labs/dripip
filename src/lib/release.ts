import * as Semver from './semver'
import * as Git from './git'
import { findIndexFromEnd, last } from './utils'

/**
 * Get the previous stable and commits since then. If there is no previous
 * stable then commits are counted from start of history.
 */
export async function getCurrentSeries(git: Git.Simple): Promise<Series> {
  return getLog(git).then(buildSeries)
}

/**
 * Get the latest stable and all subsequent commits. If no stable then all
 * commits (edge-case, new project) since-inclusive initial commit. The returned
 * tuple contains the  stable commit partitioned from the subsequent commits.
 * The stable commit may be null.
 */
async function getLog(git: Git.Simple): Promise<SeriesLog> {
  const previousStableCommit = await findLatestStable(git)
  const commits = await Git.log(git, { since: previousStableCommit })
  return previousStableCommit
    ? [commits.shift() as Git.LogEntry, commits]
    : [null, commits]
}

export type Commit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null | Semver.StableVer
    preview: null | Semver.PreviewVer
  }
}

export type StableCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: Semver.StableVer
    preview: null
  }
}

export type PreviewCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null
    preview: Semver.PreviewVer
  }
}

export type MaybePreviewCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null
    preview: null | Semver.PreviewVer
  }
}

export type UnreleasedCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null
    preview: null
  }
}

export type Series = {
  previousStable: null | StableCommit
  commitsSinceStable: MaybePreviewCommit[]
  previousPreview: null | PreviewCommit
  commitsSincePreview: UnreleasedCommit[]
  current: Commit
}

export type SeriesLog = [null | Git.LogEntry, Git.LogEntry[]]

/**
 * Build structured series data from a raw series log.
 */
export function buildSeries([previousStable, commitsSince]: SeriesLog): Series {
  if (previousStable === null && commitsSince.length === 0) {
    throw new Error(
      `Cannot build release series with given data. There is no previous stable release and no commits since. This would indicate an unexpected error or working with a git repo that has zero commits. The latter should be guarded by upstream checks. Therefore this is bad. There must be a bug.`
    )
  }

  const commitsSinceStable = commitsSince.map(c => {
    const { message, sha } = c
    return {
      message,
      sha,
      nonReleaseTags: c.tags.filter(isUnknownTag),
      releases: {
        stable: Semver.parse(c.tags.find(isStableTag) ?? ''),
        preview: Semver.parsePreview(c.tags.find(isPreviewTag) ?? ''),
      },
    } as MaybePreviewCommit
  })

  const prevPreviewI = findIndexFromEnd(
    commitsSinceStable,
    c => c.releases.preview !== null
  )

  let previousPreview: null | PreviewCommit = null
  let commitsSincePreview: UnreleasedCommit[] = []
  if (prevPreviewI !== -1) {
    previousPreview = commitsSinceStable[prevPreviewI]! as PreviewCommit
    commitsSincePreview = commitsSinceStable.slice(
      prevPreviewI + 1
    ) as UnreleasedCommit[]
  }

  const previousStable2 =
    previousStable === null
      ? null
      : ({
          sha: previousStable.sha,
          message: previousStable.message,
          nonReleaseTags: previousStable.tags.filter(isUnknownTag),
          releases: {
            stable: Semver.parse(previousStable.tags.find(isStableTag)!)!,
            preview: Semver.parsePreview(
              previousStable.tags.find(isPreviewTag) ?? ''
            ),
          },
        } as StableCommit)

  return {
    previousStable: previousStable2,
    commitsSinceStable,
    previousPreview,
    commitsSincePreview,
    // If there are no commits since stable and no stable that means we're on a
    // repo with no commit. This edge-case is ignored. It is assumed that it
    // will be validated for before calling this function.
    current: last(commitsSinceStable) ?? previousStable2!,
  }
}

function isUnknownTag(tag: string): boolean {
  return Semver.parse(tag) === null
}

function isStableTag(tag: string): boolean {
  const v = Semver.parse(tag)
  if (v === null) return false
  return Semver.isStable(v)
}

function isPreviewTag(tag: string): boolean {
  const v = Semver.parse(tag)
  if (v === null) return false
  return Semver.isPreview(v)
}

//
// Private helpers
//

async function findLatestStable(git: Git.Simple): Promise<null | string> {
  return Git.findTag(git, { matcher: isStableTag })
}

// const zeroReleases = {
//   stable: null,
//   preview: null,
// }

/**
 * Get the releases at the given commit.
 */
// export async function getReleasesAtCommit(
//   sha: string
// ): Promise<CommitReleases> {
//   const git = createGit()
//   const tags = await Git.gitGetTags(git, { ref: sha })
//   if (isEmpty(tags)) return zeroReleases
//   const parsedTags = groupByProp(tags.map(parseTag), 'type')
//   const stableTags = parsedTags.stable_release ?? []
//   const previewtags = parsedTags.pre_release ?? []

//   const invariantViolations = []
//   if (stableTags.length > 1) {
//     invariantViolations.push(
//       `Multiple stable releases found but there should only be 0 or 1: ${stableTags}`
//     )
//   }
//   if (previewtags.length > 1) {
//     invariantViolations.push(
//       `Multiple preview releases found but there should only be 0 or 1: ${stableTags}`
//     )
//   }
//   if (invariantViolations.length > 0) {
//     throw new Error(
//       `While getting the pre-existing releases at commit ${sha} the following invariant violations were found:\n\n${invariantViolations
//         .map(x => `    - ${x}`)
//         .join('\n')}`
//     )
//   }

//   return {
//     stable: stableTags[0]
//       ? {
//           type: 'stable',
//           version: stableTags[0].value,
//           sha,
//         }
//       : null,
//     preview: previewtags[0]
//       ? {
//           type: 'preview',
//           version: previewtags[0].value,
//           sha,
//           buildNum: previewtags[0].value.prerelease[1] as number,
//         }
//       : null,
//   }
// }

// export type ParsedTag =
//   | { type: 'unknown'; value: string }
//   | { type: 'stable_release'; value: SemVer.SemVer }
//   | { type: 'pre_release'; value: SemVer.SemVer }

// export function parseTag(rawTag: string): ParsedTag {
//   const semverParseResult = SemVer.parse(rawTag)
//   if (semverParseResult !== null) {
//     if (semverParseResult.prerelease.length > 0) {
//       return { type: 'pre_release', value: semverParseResult } as const
//     } else {
//       return { type: 'stable_release', value: semverParseResult } as const
//     }
//   } else {
//     return { type: 'unknown', value: rawTag } as const
//   }
// }
