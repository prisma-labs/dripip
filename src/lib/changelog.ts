/**
 * This module deals with building changelogs from a serieses.
 */
import Chalk from 'chalk'
import { stripIndents } from 'common-tags'
import { Commit, Series } from '../utils/release'
import { casesHandled } from './utils'

type ChangeLog = {
  /**
   * Any commit marked as breaking.
   */
  breaking: {
    label: string
    commits: Commit[]
  }
  /**
   * Feat type commits.
   */
  features: {
    label: string
    commits: Commit[]
  }
  /**
   * Fix type commits.
   */
  fixes: {
    label: string
    commits: Commit[]
  }
  /**
   * Any type other than feat fix and chore.
   */
  improvements: {
    label: string
    commits: Commit[]
  }
  /**
   * Commits that are intended to be ignored.
   */
  chores: {
    label: string
    commits: Commit[]
  }
  /**
   * Commit mistakes, non-conforming tools, who knows. Shouldn't happen but can.
   */
  unspecified: {
    label: string
    commits: Commit[]
  }
}

export function empty(): ChangeLog {
  return {
    breaking: { commits: [], label: 'BREAKING CHANGES' },
    features: { commits: [], label: 'Features' },
    chores: { commits: [], label: 'Chores' },
    fixes: { commits: [], label: 'Fixes' },
    improvements: { commits: [], label: 'Improvements' },
    unspecified: { commits: [], label: 'Unspecified Changes' },
  }
}

function organize(series: Series): ChangeLog {
  const log = empty()

  for (const c of series.commitsInNextStable) {
    if (c.message.parsed === null) {
      log.unspecified.commits.push(c)
      continue
    }

    const cp = c.message.parsed

    // breaking changes are collected as a group in addition to by type.
    if (cp.breakingChange) {
      log.breaking.commits.push(c)
    }

    if (cp.typeKind === 'feat') {
      log.features.commits.push(c)
    } else if (cp.typeKind === 'fix') {
      log.fixes.commits.push(c)
    } else if (cp.typeKind === 'chore') {
      log.chores.commits.push(c)
    } else if (cp.typeKind === 'other') {
      log.improvements.commits.push(c)
    } else {
      casesHandled(cp.typeKind)
    }
  }

  return log
}

export function render(
  series: Series,
  opts: { type: 'plain' | 'markdown' }
): string {
  if (opts.type === 'markdown') return Markdown.render(organize(series))
  if (opts.type === 'plain') return Plain.render(organize(series))
  casesHandled(opts.type)
}

namespace Plain {
  export function render(log: ChangeLog): string {
    const order: (keyof Omit<ChangeLog, 'unspecified'>)[] = [
      'breaking',
      'features',
      'fixes',
      'improvements',
      'chores',
    ]

    const doc = order
      .filter(sectionName => {
        return log[sectionName].commits.length > 0
      })
      .map(sectionName => {
        if (sectionName === 'breaking') {
          return (
            sectionTitle(log[sectionName].label) +
            '\n\n' +
            sectionCommits(log[sectionName].commits, { breaking: false }) +
            '\n'
          )
        }

        if (sectionName === 'improvements') {
          return (
            sectionTitle(log[sectionName].label) +
            '\n\n' +
            sectionCommits(log[sectionName].commits, { type: true }) +
            '\n'
          )
        }

        return (
          sectionTitle(log[sectionName].label) +
          '\n\n' +
          sectionCommits(log[sectionName].commits) +
          '\n'
        )
      })

    if (log.unspecified.commits.length) {
      doc.push(
        sectionTitle(log.unspecified.label) +
          '\n\n' +
          '  ' +
          log.unspecified.commits
            .map(c => `${Chalk.gray(shortSha(c))} ${c.message.raw}`)
            .join('\n  ') +
          '\n'
      )
    }

    return doc.join('\n')
  }

  function sectionCommits(cs: Commit[], opts?: CommitRenderOpts): string {
    return cs.map(c => sectionCommit(c, opts)).join('\n')
  }

  function sectionTitle(title: string): string {
    return Chalk.magenta(title)
  }

  type CommitRenderOpts = { type?: boolean; breaking?: boolean }

  function sectionCommit(c: Commit, opts?: CommitRenderOpts): string {
    const sha = Chalk.gray(shortSha(c))
    const type = opts?.type === true ? ' ' + c.message.parsed!.type + ':' : ''
    const description = ' ' + c.message.parsed!.description
    const breaking =
      opts?.breaking === false
        ? ''
        : c.message.parsed!.breakingChange
        ? Chalk.red(' (breaking)')
        : ''
    return `  ${sha}${breaking}${type}${description}`
  }
}

namespace Markdown {
  export function render(log: ChangeLog): string {
    const order: (keyof Omit<ChangeLog, 'unspecified'>)[] = [
      'breaking',
      'features',
      'fixes',
      'improvements',
      'chores',
    ]

    const doc = order
      .filter(sectionName => {
        return log[sectionName].commits.length > 0
      })
      .map(sectionName => {
        if (sectionName === 'breaking') {
          return (
            stripIndents`
            ${sectionTitle(log[sectionName].label)}
  
            ${log[sectionName].commits
              .map(c => sectionCommit(c, { breaking: false }))
              .join('\n')}
          ` + '\n'
          )
        }

        if (sectionName === 'improvements') {
          return (
            stripIndents`
            ${sectionTitle(log[sectionName].label)}
  
            ${log[sectionName].commits
              .map(c => sectionCommit(c, { type: true }))
              .join('\n')}
          ` + '\n'
          )
        }

        return (
          stripIndents`
          ${sectionTitle(log[sectionName].label)}

          ${sectionCommits(log[sectionName].commits)}
        ` + '\n'
        )
      })

    if (log.unspecified.commits.length) {
      doc.push(
        stripIndents`
        ${sectionTitle(log.unspecified.label)}

        - ${log.unspecified.commits
          .map(c => `${shortSha(c)} ${c.message.raw}`)
          .join('\n- ')}
      ` + '\n'
      )
    }

    return doc.join('\n')
  }
  function sectionCommits(cs: Commit[]): string {
    return cs.map(c => sectionCommit(c)).join('\n')
  }

  function sectionTitle(title: string): string {
    return `#### ${title}`
  }

  function sectionCommit(
    c: Commit,
    opts?: { type?: boolean; breaking?: boolean }
  ): string {
    const sha = shortSha(c)
    const type = opts?.type === true ? ' ' + c.message.parsed!.type + ':' : ''
    const description = ' ' + c.message.parsed!.description
    const breaking =
      opts?.breaking === false
        ? ''
        : c.message.parsed!.breakingChange
        ? ' (breaking)'
        : ''
    return `- ${sha}${breaking}${type}${description}`
  }
}

function shortSha(c: Commit): string {
  return c.sha.slice(0, 7)
}
