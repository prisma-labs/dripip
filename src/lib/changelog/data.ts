/**
 * This module deals with building changelogs from multiple series.
 */
import { Commit } from '../../utils/release'

export type Changelog = {
  /**
   * Any commit marked as breaking.
   */
  breaking: {
    label: string
    commits: Commit[]
  }
  /**
   * Feat type commits.
   */
  features: {
    label: string
    commits: Commit[]
  }
  /**
   * Fix type commits.
   */
  fixes: {
    label: string
    commits: Commit[]
  }
  /**
   * Any type other than feat fix and chore.
   */
  improvements: {
    label: string
    commits: Commit[]
  }
  /**
   * Commits that are intended to be ignored.
   */
  chores: {
    label: string
    commits: Commit[]
  }
  /**
   * Commit mistakes, non-conforming tools, who knows. Shouldn't happen but can.
   */
  unspecified: {
    label: string
    commits: Commit[]
  }
}

/**
 * Crate an empty changelog.
 */
export function empty(): Changelog {
  return {
    breaking: { commits: [], label: 'BREAKING CHANGES' },
    features: { commits: [], label: 'Features' },
    chores: { commits: [], label: 'Chores' },
    fixes: { commits: [], label: 'Fixes' },
    improvements: { commits: [], label: 'Improvements' },
    unspecified: { commits: [], label: 'Unspecified Changes' },
  }
}
