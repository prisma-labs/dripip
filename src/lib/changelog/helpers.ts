import { Series } from '../../utils/release'
import { casesHandled } from '../utils'
import { Changelog, empty } from './data'
import { Markdown, Terminal } from './renderers'

export function renderFromSeries(series: Series, options: RenderOptions): string {
  return render(fromSeries(series), options)
}

export type RenderOptions = {
  as: `plain` | `markdown`
}

/**
 * Render a changelog into a string using the chosen renderer (Markdown, Terminal, etc.).
 */
export function render(changelog: Changelog, opts: RenderOptions): string {
  if (opts.as === `markdown`) return Markdown.render(changelog)
  if (opts.as === `plain`) return Terminal.render(changelog)
  casesHandled(opts.as)
}

/**
 * Transform a series into a changelog.
 */
export function fromSeries(series: Series): Changelog {
  const log = empty()

  for (const c of series.commitsInNextStable) {
    if (c.message.parsed === null) {
      log.unspecified.commits.push(c)
      continue
    }

    const cp = c.message.parsed

    // breaking changes are collected as a group in addition to by type.
    if (cp.breakingChange) {
      log.breaking.commits.push(c)
    }

    if (cp.typeKind === `feat`) {
      log.features.commits.push(c)
    } else if (cp.typeKind === `fix`) {
      log.fixes.commits.push(c)
    } else if (cp.typeKind === `chore`) {
      log.chores.commits.push(c)
    } else if (cp.typeKind === `other`) {
      log.improvements.commits.push(c)
    } else {
      casesHandled(cp.typeKind)
    }
  }

  return log
}
