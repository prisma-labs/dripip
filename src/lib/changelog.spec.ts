// Disable colour from test output
// https://github.com/chalk/chalk#chalklevel
process.env.FORCE_COLOR = '0'

import * as Release from '../utils/release'
import * as Changelog from './changelog'

describe('.render()', () => {
  const mockChangeLog = [
    { message: 'chore: b' },
    { message: 'fix: b' },
    { message: 'feat: b' },
    { message: 'refactor: a' },
    { message: 'non conforming commit b' },
    { message: 'chore: a' },
    { message: 'feat: a\n\nBREAKING CHANGE:\ntoto' },
    { message: 'perf: a' },
    { message: 'fix: a' },
    { message: 'non conforming commit a' },
    { message: 'feat: blah', version: '0.1.0' },
  ]
  describe('option markdown', () => {
    const render = (...logs: MockLogs) => {
      return Changelog.renderChangelog(series(...logs), { as: 'markdown' })
    }
    it('renders release notes for the current release series', () => {
      const notes = render(...mockChangeLog)
      expect(notes).toMatchSnapshot()
    })
    it('shows breaking commits twice, once in breaking section and once in its native section', () => {
      const notes = render({ message: 'a: foo\n\nBREAKING CHANGE:\nbar' })
      expect(notes).toMatch(/BREAKING CHANGES\n\n.*foo/)
      expect(notes).toMatch(/Improvements\n\n.*foo/)
    })
    it('shows breaking label if commit breaking except within breaking section', () => {
      const notes = render({ message: 'a: foo\n\nBREAKING CHANGE:\nbar' })
      expect(notes).toMatch(/Improvements\n\n.*(breaking).*foo/)
    })
    it('hides unknown section if no unknown commits', () => {
      const notes = render({ message: 'a: b' })
      expect(notes).not.toMatch(Changelog.empty().unspecified.label)
    })
    it('improvements section prefixes commits with their type', () => {
      const notes = render({ message: 'a: foo' })
      expect(notes).toMatch(/a: foo/)
    })
  })

  describe('plain', () => {
    const render = (...logs: MockLogs) => {
      return Changelog.renderChangelog(series(...logs), { as: 'plain' })
    }

    it('renders release notes for the current release series', () => {
      const notes = render(...mockChangeLog)
      expect(notes).toMatchSnapshot()
    })
  })
})

type MockLogs = { message: string; version?: string }[]

function series(...logs: MockLogs): Release.Series {
  return Release.fromLogs(
    logs.map(({ message, version }) => ({
      message,
      tags: version ? [version] : [],
      sha: 'shasha#',
    }))
  )
}
