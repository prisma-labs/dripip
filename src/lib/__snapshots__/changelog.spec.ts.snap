// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.render() option markdown renders release notes for the current release series 1`] = `
"#### BREAKING CHANGES

- shasha# a

#### Features

- shasha# b
- shasha# (breaking) a

#### Fixes

- shasha# b
- shasha# a

#### Improvements

- shasha# refactor: a
- shasha# perf: a

#### Chores

- shasha# b
- shasha# a

#### Unspecified Changes

- shasha# non conforming commit b
- shasha# non conforming commit a
"
`;

exports[`.render() plain renders release notes for the current release series 1`] = `
"[35mBREAKING CHANGES[39m

  [90mshasha#[39m a

[35mFeatures[39m

  [90mshasha#[39m b
  [90mshasha#[39m[31m (breaking)[39m a

[35mFixes[39m

  [90mshasha#[39m b
  [90mshasha#[39m a

[35mImprovements[39m

  [90mshasha#[39m refactor: a
  [90mshasha#[39m perf: a

[35mChores[39m

  [90mshasha#[39m b
  [90mshasha#[39m a

[35mUnspecified Changes[39m

  [90mshasha#[39m non conforming commit b
  [90mshasha#[39m non conforming commit a
"
`;
